### ISSUE TITLE
1.1.0-pre --unified builds (mingw*) take 35-40% longer with double the log output

### ISSUE THREAD

**Issue by vszakats** [openssl/openssl#750]
Since about 8-9 days ago, builds became more lengthy (in time) and emit double the log output. This appears to be because now `gcc` is invoked for each source twice:
```
gcc -DOPENSSL_THREADS -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSL_PIC -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DRMD160_ASM -DAES_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines\""  -D_MT -DL_ENDIAN -DWIN32_LEAN_AND_MEAN -DUNICODE -D_UNICODE -m32 -Wall -O3 -fomit-frame-pointer -DWHIRLPOOL_ASM -fno-ident -D_WINDLL -Iinclude -I. -Icrypto/include -MM -MF crypto/camellia/cmll_ofb.d -MQ crypto/camellia/cmll_ofb.o crypto/camellia/cmll_ofb.c
touch crypto/camellia/cmll_ofb.d
gcc -DOPENSSL_THREADS -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSL_PIC -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DRMD160_ASM -DAES_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines\""  -D_MT -DL_ENDIAN -DWIN32_LEAN_AND_MEAN -DUNICODE -D_UNICODE -m32 -Wall -O3 -fomit-frame-pointer -DWHIRLPOOL_ASM -fno-ident -D_WINDLL -Iinclude -I. -Icrypto/include -c -o crypto/camellia/cmll_ofb.o crypto/camellia/cmll_ofb.c
```

Example:
https://ci.appveyor.com/project/vsz/harbour-deps/build/1.0.660
using current `master` `HEAD`.

Example for the previous iteration:
https://ci.appveyor.com/project/vsz/harbour-deps/build/1.0.658
using `master` d1776fdecd6a8677a281a85498a74a8480739689

These commits may be the reason:
* "Automatic 'make depend' for the unified build scheme"
c058fcd79877530f18a288664ea71def012c405a
* "Don't treat .d (depend) files separately from object files"
88297284ad14a233430cb5140d368edf17dacf7a

Running a `make depend` appears to be a no-op now, so it didn't help to avoid the `.d` pass.

This is not an _error_, but the old logs were easier to work with and the 2.5 minutes shorter builds (per target) were a definite _nice thing_ to have. I'm not sure what the extra `.d` phase does now, but for clean builds with no intent of further incremental passes, _maybe_ it would be feasible to drop those. Or maybe it'd be possible to merge these two separate calls into one? (on Windows, the overhead of each `gcc` invocation is quite high unfortunately)

/cc @levitte

**Comment by vszakats** [comment]
Updated the title. To be correct he _total_ build time increased from 20 minutes to 25 minutes, of which the OpenSSL revision accounts for the 5 minutes increase, but there are other build tasks included in the total time worth about 7 minutes, so the time increase as percentage for the OpenSSL tasks alone is more around 35-40%.

**Comment by levitte** [comment]
The construction that you're talking about is to have automatic dependency generation.  The little `.d` files are dependency information for the corresponding object file, and `make depend` collects them all and appends them to `Makefile`.

The reason that we treat the building of the `.d` files is that they don't quite depend on the same files as the object files, so regenerating them at the same time as building the object files could ultimately cause unecessary rebuilds of `Makefile`.  However, I'm currently thinking there's a different way to go about it, we could actually generate dependencies into temporary files with the object file building call, and then rename to the `.d` file *if they differ*.  I'll have a go at that,

**Comment by vszakats** [comment]
Thank you!

In above builds, there are only three build steps: a (manual) clean, a `./Configure` and a `make`. I was assuming `./Configure` creates all the required `Makefile`s and `make` just executes them. Since it's a clean (= non-incremental) build, all the make steps will need to be executed unconditionally, which is typical in CI scenarios. AFAIU the `.d` files have a use when one wants to do an incremental build without the burden of maintaining header and other cross source (non .c &rarr; .o) dependencies manually. I'm assuming this is what's being done now as part of the normal build process. Not sure what is best way to support both make scenarios in an optimal fashion.

Looking forward to test any new commit in this regard.

**Comment by levitte** [comment]
Just worked on it, came up with this change.  Feel free to try it out.

```
diff --git a/Configurations/unix-Makefile.tmpl b/Configurations/unix-Makefile.tmpl
index b9fc5f8..cff0023 100644
--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -847,11 +847,12 @@ $obj$objext: $obj$depext
 EOF
       }
       return <<"EOF";
-$obj$depext: $deps
-       \$(CC) \$(CFLAGS) $ecflags$incs -MM -MF \$\@ -MQ $obj$objext $srcs
-       touch \$\@
-$obj$objext: $obj$depext
-       \$(CC) \$(CFLAGS) $ecflags$incs -c -o \$\@ $srcs
+$obj$objext: $deps
+       \$(CC) \$(CFLAGS) $ecflags$incs -MMD -MF $obj$depext.tmp -MT \$\@ -c -o \$\@ $srcs
+       \@touch $obj$depext.tmp
+       \@if ! cmp $obj$depext.tmp $obj$depext > /dev/null 2> /dev/null; then \\
+               mv $obj$depext.tmp $obj$depext; \\
+       fi
 EOF
   }
   # On Unix, we build shlibs from static libs, so we're ignoring the
```

**Comment by levitte** [comment]
That was just merged, commit 340da949ccde13f6506177e327ccc21f49f56d3b

Closing this now

**Comment by vszakats** [comment]
Thank you very much, build output back to its short form. (build times should follow)
